### PR TITLE
[Merged by Bors] - feat(NumberTheory/Padics/PadicVal): padicValNat_factorial

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 1]
```diff
@@ -549,6 +556,16 @@ theorem padicValNat_factorial_mul {p : ℕ} (n : ℕ) (hp : p.Prime):
       padicValNat_def' (Nat.Prime.ne_one hp) <| factorial_pos n]
   exact Prime.multiplicity_factorial_mul hp
 
+/-- **Legendre's Theorem**
+
+The `p`-adic valuation of `n!` is the sum of the quotients `n / p ^ i`. This sum is expressed
+over the finset `Ico 1 b` where `b` is any bound greater than `log p n`. -/
+theorem multiplicity_factorial {p : ℕ} [hp : Fact p.Prime] :
+    ∀ {n b : ℕ}, log p n < b → (padicValNat p (n !)) = (∑ i in Finset.Ico 1 b, n / p ^ i : ℕ) :=
```
Comment (ocfnash) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:564]: Surely the statement should be:
```suggestion
theorem padicValNat_factorial {p : ℕ} [hp : Fact p.Prime] {n : ℕ} :
    padicValNat p (n !) = ∑ i in Finset.Icc 1 (log p n), n / p ^ i :=
```

File: Mathlib/Data/Nat/Multiplicity.lean [hunk 2]
```diff
@@ -26,7 +26,8 @@ coefficients.
 ## Multiplicity calculations
 
 * `Nat.Prime.multiplicity_factorial`: Legendre's Theorem. The multiplicity of `p` in `n!` is
-  `n/p + ... + n/p^b` for any `b` such that `n/p^(b + 1) = 0`.
+  `n/p + ... + n/p^b` for any `b` such that `n/p^(b + 1) = 0`. See `padicValNat_factorial` for the
+  for the corresponding theorem on `p`-adic valuations.
```
Comment (ocfnash) [inline Mathlib/Data/Nat/Multiplicity.lean:30]:
```suggestion
  for this result stated in the language of `p`-adic valuations.
```

**Review by ocfnash — CHANGES_REQUESTED** [review]
Thanks!

Only minor suggestions but I think worth doing.

File: Mathlib/Data/Nat/Multiplicity.lean (hunk 2 above)
Comment (mo271) [reply Mathlib/Data/Nat/Multiplicity.lean]: Great! I also edited the similar sentence in the `padicValNat` file.

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 3]
```diff
@@ -549,6 +556,15 @@ theorem padicValNat_factorial_mul {p : ℕ} (n : ℕ) (hp : p.Prime):
       padicValNat_def' (Nat.Prime.ne_one hp) <| factorial_pos n]
   exact Prime.multiplicity_factorial_mul hp
 
+/-- **Legendre's Theorem**
+
+The `p`-adic valuation of `n!` is the sum of the quotients `n / p ^ i`. This sum is expressed
+over the finset `Ico 1 b` where `b` is any bound greater than `log p n`. -/
+theorem padicValNat_factorial {p n b : ℕ} [hp : Fact p.Prime] (hnb : log p n < b) :
+    (padicValNat p (n !)) = (∑ i in Finset.Ico 1 b, n / p ^ i : ℕ) :=
```
Comment (ocfnash) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:564]: The extra brackets can go:
```suggestion
    padicValNat p (n !) = ∑ i in Finset.Ico 1 b, n / p ^ i :=
```

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**